media: adi-fb: Add ADI FB AXI driver

btogorean opened the pull request:
This PR propose a new driver for control of AXI Frame Buffer HDL core.
More details here: https://wiki.analog.com/resources/fpga/docs/axi_dmac in section FrameLock Synchronization.

This is the preliminary, standalone version of driver as required by Mathworks.
Possible future integration with V4L2 framework.

nunojsa commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,296 @@
+	if (!frm_buff) {
+		ret = ENOMEM;
+		goto err;
+	}
```
I guess you could just return `-ENOMEM`?

btogorean replied:
:) reminiscence from more complex error handling.
But I agree that how it its now it should return right away.

nunojsa commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,296 @@
+	/* Get reserved memory region from Device-tree */
+	np = of_parse_phandle(pdev->dev.of_node, "memory-region", 0);
+	if (!np)
+		dev_err(&pdev->dev, "No %s specified\n", "memory-region");
```
Why not returning right way? If `of_address_to_resource()` fails, i think there's no point in going to `of_address_to_resource()`, right?

btogorean replied:
agree

commodo replied:
is there a `fwnode_` variant to this function ?
it would be interesting to have this to also maybe support ACPI in some future;

nunojsa commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,296 @@
+static int frame_buffer_remove(struct platform_device *pdev)
+{
+	struct frame_buffer *frm_buff = platform_get_drvdata(pdev);
+
```
Most likely, I'm missing something but shouldn't you call `platform_set_drvdata()` in the `probe()` function?

btogorean replied:
ups. that's quite a problem
thank you for detecting it

nunojsa commented on Documentation/devicetree/bindings/media/adi,adi-fb.yaml:
```diff
@@ -0,0 +1,98 @@
+            reserved: buffer@0 {
+                    no-map;
+                    reg = <0x1C000000 0x2000000>;
+                    };
```
nitpick, to much indentation?

btogorean replied:
if you're referring to last curly bracket I agree

nunojsa replied:
exactly that one :smile:

btogorean commented:
Changes in V2:
- removed "goto" and return directly
- call platform_set_drvdata() in probe function
- fix indentation in devicetree bindings

commodo commented on drivers/media/platform/Kconfig:
```diff
@@ -526,6 +526,13 @@ config VIDEO_TI_SC
 config VIDEO_TI_CSC
 	tristate
 
+config VIDEO_FRAME_BUFFER
```
maybe a more generic name would be needed here;
otherwise it has changes of coliding with other symbols
VIDEO_FRAME_BUFFER -> ADI_VIDEO_FRAME_BUFFER  or something like that

btogorean replied:
ack

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,295 @@
+};
+
+static u32 adi_fb_reg_read(void __iomem *base_addr,
+			   u32 addr)
```
this would fit on one line; there is no need to split this on multiple lines

btogorean replied:
ack

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,295 @@
+	return ioread32(base_addr + addr);
+}
+
+static void adi_fb_reg_write(void __iomem *base_addr,
```
this also fits on one line

btogorean replied:
ack

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,295 @@
+		       DMAC_CTRL_ENABLE);           // enable DMAC
+	adi_fb_reg_write(buff->hdl_subdev.rx_dma_regs, DMAC_REG_FLAGS,
+			 (
+				 DMAC_FLAGS_FLOCK  |
```
this looks like it could be made into a macro that includes all these values;
something like 
#define DMAC_DEFAULT_FLAGS (DMAC_FLAGS_FLOCK | DMAC_FLAGS_CYCLIC | DMAC_FLAGS_TLAST)

btogorean replied:
agree

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,295 @@
+	adi_fb_reg_write(buff->hdl_subdev.tx_dma_regs,
+			 DMAC_REG_FRAME_LOCK_CONFIG,
+			 (buff->num_frames |            // FLOCK_NUMFRAMES
+			  (1 << 9) |
```
why `1 << 9` ?

btogorean replied:
Rx DMA should have FLOCK_WAIT_WRITER (1<<9) flag set.
Will move this in a macro

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,295 @@
+		       DMAC_TRANSFER_SUBMIT);           // submit transfer
+
+	// Init TX DMAC
+	adi_fb_reg_set(buff->hdl_subdev.tx_dma_regs, DMAC_REG_CTRL,
```
the TX & RX initializations looks fairly similar;
I guess they could be grouped into a single function with some minor variations that can handle both;

btogorean replied:
agree

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,295 @@
+	}
+
+	/* Get operating mode */
+	if (device_property_read_u32(&pdev->dev, "flock,mode",
```
for many of these properties, it's a good idea to use `adi,` prefix;  so something like `adi,flock-mode` 
this is to make it clearer that it's an ADI prefix; it's usually recommended when driver is upstream

btogorean replied:
naming of this driver was not clear but will stick to adi for now and if Mathworks don't agree it I will change according

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,295 @@
+		frm_buff->num_frames = 3;
+		dev_info(&pdev->dev, "No frames number specified. Using %d\n",
+			 frm_buff->num_frames);
+
```
nitpick: extra line

btogorean replied:
ack

btogorean commented:
Changes in V3:

- indentation/spacing remarks fixed

- added adi prefix to devicetree properties

- adi_fb_init handle both tx/rx

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,263 @@
+#include <media/media-device.h>
+
+/* DMA defines */
+#define DMAC_REG_CTRL 0x400
```
style-wise, it is preferred to indent the values of macros on the same column (using tabs)
so, something like
```
#define DMAC_REG_START_TRANSFER    0x408
#define DMAC_REG_FLAGS                          0x40c
#define DMAC_REG_DEST_ADDRESS        0x41
```

btogorean replied:
ack

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,263 @@
+#define DMAC_REG_FRAME_LOCK_CONFIG 0x454
+#define DMAC_REG_FRAME_LOCK_STRIDE 0x458
+
+#define DMAC_CTRL_ENABLE (1 << 0)
```
the BIT(x) macro could be used here

btogorean replied:
ack

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,263 @@
+
+static void adi_fb_reg_set(void __iomem *base_addr, u32 addr, u32 set)
+{
+	adi_fb_reg_write(base_addr, addr, adi_fb_reg_read(base_addr,
```
this looks a bit funky style-wise
maybe
```
	adi_fb_reg_write(base_addr, addr,
                                    adi_fb_reg_read(base_addr, addr) | set);
```
or add a new line:
```
set |= adi_fb_reg_read(base_addr, addr);
adi_fb_reg_write(base_addr, addr, set);
```

same comment for `adi_fb_reg_clr()`

btogorean replied:
ack

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,263 @@
+	adi_fb_reg_write(base_addr, DMAC_REG_FLAGS, DMAC_DEFAULT_FLAGS);
+
+	adi_fb_reg_write(base_addr,
+			 dma_dir ? DMAC_REG_DEST_ADDRESS : DMAC_REG_SRC_ADDRESS,
```
why not just assign `DMAC_REG_DEST_ADDRESS / DMAC_REG_SRC_ADDRESS` as a param to `adi_fb_init(buff, reg)` ?

commodo replied:
oh wait; i forgot about stride;
maybe use temp variables inside the function
```
	if (dma_dir == RX_DMA) {
		base_addr = buff->hdl_subdev.rx_dma_regs;
                dir = DMAC_REG_DEST_ADDRESS;
                stride = DMAC_REG_DEST_STRIDE;
	} else {
		base_addr = buff->hdl_subdev.tx_dma_regs;
                dir = DMAC_REG_SRC_ADDRESS;
                stride = DMAC_REG_SRC_STRIDE;
        }
```

btogorean replied:
if ternary is not liked by upstream guys agree

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,263 @@
+
+	adi_fb_reg_write(base_addr,
+			 DMAC_REG_FRAME_LOCK_CONFIG,
+			 (buff->num_frames |            // FLOCK_NUMFRAMES
```
maybe assign this to a variable
it seems to be doing too much inside the param call

btogorean replied:
agree

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,263 @@
+
+static struct platform_driver frame_buffer_driver = {
+	.driver			= {
+		.name		= "frame-buffer",
```
`frame-buffer` -> `adi-frame-buffer` ?

btogorean replied:
oups. missed this one. thanks

stefpopa commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,263 @@
+	u32 num_frames;
+	u32 mode;
+	u32 distance;
+	//resolution[0] width, resolution[1] height
```
Please use C style comments; check this everywhere in your code

btogorean replied:
ack

stefpopa commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,263 @@
+static void adi_fb_reg_clr(void __iomem *base_addr, u32 addr, u32 clr)
+{
+	adi_fb_reg_write(base_addr, addr, adi_fb_reg_read(base_addr,
+							  addr) & ~clr);
```
nitpick: would it fit if we moved the third parameter on a different line?
```
adi_fb_reg_write(base_addr, addr,
                 adi_fb_reg_read(base_addr, addr) & ~clr);
```

btogorean replied:
it fits :) will update

stefpopa commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,263 @@
+		frm_buff->num_frames = 3;
+		dev_info(&pdev->dev, "No frames number specified. Using %d\n",
+			 frm_buff->num_frames);
+	}
```
I know that this probably works, but to me it is not that obvious what happens here.
I would rewrite this:
```
	ret = device_property_read_u32(&pdev->dev, "adi,flock-frm-buf-nr", &tmp);
	if (ret) {
		dev_info(&pdev->dev, "No frames number specified");
		frm_buff->num_frames = 3;
	} else {
		frm_buff->num_frames = tmp;
	}
```
However, I do not feel strongly about this, so feel free to disregard.

btogorean replied:
you are right about clarity. will update. thank you

stefpopa commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,263 @@
+{
+	struct frame_buffer *frm_buff = platform_get_drvdata(pdev);
+
+	media_device_unregister(&frm_buff->media_dev);
```
Would it be an idea to use `devm_add_action_or_reset()`? This way we could get rid of the `frame_buffer_remove()` altogether.  Here's an example: https://elixir.bootlin.com/linux/v5.1.1/source/drivers/iio/adc/ad7768-1.c#L576

btogorean replied:
good suggestion. thank you

stefpopa commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,263 @@
+
+	u32 num_frames;
+	u32 mode;
+	u32 distance;
```
nitpick: you can probably fit multiple u32 variables on a single line

btogorean replied:
ack

commodo replied:
yep; each u32 on 1 line is fine

btogorean commented:
Changes in V4:

- fixed indentation/C style comments
- dropped use of ternary operator
- removed frame_buffer_remove() function and replaced with  devm_add_action_or_reset()
- changes for clarity to device_property_read_u32() returned response handling

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,270 @@
+		return -ENODEV;
+	}
+
+	ret = of_address_to_resource(np, 0, &frm_buff->video_ram_buf);
```
same for this one; is there a `fwnode_` variant ?

btogorean replied:
I don't think this memory reserve makes sense in ACPI context

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,270 @@
+	}
+
+	ret = devm_add_action_or_reset(&pdev->dev,
+		(void(*)(void *))media_device_unregister, &frm_buff->media_dev);
```
casting functions is bad practice; it can hide various issues;
maybe add a function that wraps this; then, later, the compiler can catch stuff

btogorean replied:
ack. wrapper will be created

commodo commented:
rest looks good

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,270 @@
+#include <media/media-device.h>
+
+/* DMA defines */
+#define DMAC_REG_CTRL				0x400
```
can you take a look at using `linux/fpga/adi-axi-common.h` ?
some common regs are defined there
and are common for multiple AXI IP cores we provide

btogorean replied:
ack

commodo commented on Documentation/devicetree/bindings/media/adi,adi-fb.yaml:
```diff
@@ -0,0 +1,98 @@
+properties:
+  compatible:
+    enum:
+        - adi,framebuffer-0.00.a
```
maybe let's make this `adi,axi-framebuffer-1.00.a`

btogorean replied:
ack

commodo commented on Documentation/devicetree/bindings/media/adi,adi-fb.yaml:
```diff
@@ -0,0 +1,98 @@
+examples:
+  - |
+    adi-fb {
+            compatible = "adi,framebuffer-0.00.a";
```
same here; maybe `adi,axi-framebuffer-1.00.a`

btogorean replied:
ack

commodo commented on drivers/media/platform/Kconfig:
```diff
@@ -526,6 +526,13 @@ config VIDEO_TI_SC
 config VIDEO_TI_CSC
 	tristate
 
+config ADI_VIDEO_FRAME_BUFFER
```
maybe let's maket this `ADI_AXI_VIDEO_FRAME_BUFFER` or just `AXI_VIDEO_FRAME_BUFFER`

btogorean replied:
ADI_AXI_VIDEO_FRAME_BUFFER it is to not interfere with other possible future AXI FB drivers

commodo commented on drivers/media/platform/Makefile:
```diff
@@ -56,6 +56,7 @@ obj-y					+= davinci/
 obj-$(CONFIG_VIDEO_SH_VOU)		+= sh_vou.o
 obj-$(CONFIG_VIDEO_AXI_HDMI_RX)		+= axi-hdmi-rx.o
 obj-$(CONFIG_VIDEO_IMAGEON_BRIDGE)	+= imageon-bridge.o
+obj-$(CONFIG_ADI_VIDEO_FRAME_BUFFER)	+= adi-fb.o
```
`adi-fb.o` -> `axi-fb.o`

btogorean replied:
ack

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,270 @@
+
+static struct platform_driver frame_buffer_driver = {
+	.driver			= {
+		.name		= "adi-frame-buffer",
```
`axi-frame-buffer`

btogorean replied:
ack

commodo commented on drivers/media/platform/adi-fb.c:
```diff
@@ -0,0 +1,270 @@
+}
+
+static const struct of_device_id frame_buffer_of_match[] = {
+	{ .compatible = "adi,framebuffer-0.00.a", },
```
`adi,axi-framebuffer-1.00.a`

btogorean replied:
ack

btogorean commented:
Changes in V5:

- added wrapper function for devm_add_action_or_reset()
- changed names from adi-fb* to adi-axi-fb

commodo commented:
Will merge tomorrow if no objections
